use accumulate type in BF16 gemm(include dot, mv) ref path

zhuhaozhe opened the pull request:
Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96074



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

zhuhaozhe committed:
use accmulate type in BF16 gemm(include dot, mv) ref path

[ghstack-poisoned]

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"




[ghstack-poisoned]

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -322,9 +322,34 @@ void gemm(
+   std::vector<float> float_c(n * ldc);
+   std::vector<float> float_a(k * lda);
+   std::vector<float> float_b(n * ldb);
+   for (auto& av : float_a) {
```
When tensor a or b is not contiguous, that is to say, lda != m or lab != k, can we still use `a++` or `b++` here ?

lezcano replied:
To do this, you probably want to do `a.to(at::kFloat, at::MemoryFormat::Contiguous)`, which you know for sure it's properly vectorized. 

Even then, doesn't `gemm_stub` dispatch to `gemm_core_`, which already does this via `opmath` directly in the kernel?

mingfeima commented on test/test_linalg.py:
```diff
@@ -7449,6 +7449,42 @@ def test_permute_matmul(self):
+                                                      f"expected: {expected}\n" \
+                                                      f"got: {c}\n"
+        # test matmul
+        a = torch.ones(300, 300)
```
need to add cases when a or b is not contiguous here.

zhuhaozhe replied:
Thanks for reminder, have added checks for `notrans`, `transa`, `transb`, and `transab` for different `gemm_core_` impl

lezcano commented:
Also, there are comprehensive tests that should tests these paths in `test_linalg.py` under the name of `test_matmul_small_brute_force_2d_Nd`. Have a look at them and, if you add new tests, make sure that they don't overlap with those.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



[ghstack-poisoned]

zhuhaozhe commented:
> Also, there are comprehensive tests that should tests these paths in `test_linalg.py` under the name of `test_matmul_small_brute_force_2d_Nd`. Have a look at them and, if you add new tests, make sure that they don't overlap with those.

Thanks for reminder, I checked this UT and this ut is only set on FP32 which is different with my newly added one.
My newly added one is aim to check whether `256 + 1` is taken place at fp32 level in bf16 gemm ref path since `256(bf16) + 1(bf16) = 256(bf16)`

jgong5 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1550,18 +1552,18 @@ inline void baddbmm_cpu_kernel(const Tensor& result, const Tensor& self, const T
-                r += s2[k] * m1[k][j];
+                acc_value += s2[k] * m1[k][j];
               }
             } else {
-              r *= beta;
+              acc_value *= beta;
```
You might have to handle conflict from this PR: https://github.com/pytorch/pytorch/pull/96086. cc @XiaobingSuper

zhuhaozhe replied:
Got it, will rebase this PR after https://github.com/pytorch/pytorch/pull/96086 is merged.

jgong5 commented on test/test_linalg.py:
```diff
@@ -7449,6 +7449,48 @@ def test_permute_matmul(self):
+        # and https://github.com/pytorch/pytorch/issues/83863
+        # for bf16 accumulation in gemm ref path
+        def check_correctness(fn, *args):
+                expected = fn(*args).bfloat16()
```
For ref result, is it ok to use fp32 inputs directly or do a fake quant with `bfloat16()` and `float()` for exact match with bfloat16 result?

zhuhaozhe replied:
We do not need fake quant here since we always have new tensors whose elements are `1`. I do not see acc loss for `[1.0,].bfloat16().float()`.

jgong5 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -180,18 +181,22 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
     return;
   }
 
+  using accscalar_t = at::acc_type<scalar_t, false>;
```
How about using `opmath_t` consistently?

zhuhaozhe replied:
Sounds good, have followed this way.

jgong5 commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -53,32 +53,30 @@ auto sum(int64_t N, Func f) {
-        c[j * ldc + i] += a[i + l * lda] * val;
+      const auto dot = sum(k, [&](int64_t l) -> opmath_t {
+        return static_cast<opmath_t>(a[i * lda + l]) *
+            static_cast<opmath_t>(b[j * ldb + l]);
```
What does this function assume about `b`'s transposition? Here we assume `b` is notrans, right?

zhuhaozhe replied:
Yes, it is wrong here.
But I confirmed
```        
return static_cast<opmath_t>(a[l * lda + i]) *
            static_cast<opmath_t>(b[j * ldb + l]);
```
can work for whether b is transposed or not.
`transpose` will impact the value for `ldb` which will make  `b[j * ldb + l]` works anyway.

peterbell10 replied:
> `transpose` will impact the value for `ldb` which will make `b[j * ldb + l]` works anyway.

It will be valid to read, but you'll be reading the wrong dimension.

zhuhaozhe replied:
Yes, you are right. I am reading the wrong dimension, have fixed.

lezcano replied:
Why didn't the tests catch this issue?

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano reviewed (approved):
Please avoid formatting files when submitting fixes, as it makes reviewing (and later on, digging for who last modified the code) unnecessarily difficult.

peterbell10 commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -110,35 +108,6 @@ void gemm_transa_(
 }
 
-template <typename scalar_t, typename opmath_t>
-void gemm_transb_(
```
Why are you removing these kernels entirely, rather than specializing `gemm_core` for Half/BFloat16?

zhuhaozhe replied:
I thought `transb_` and `notrans_` can be unified into one kernel so I removed it. But as you found before, I was read the wrong dimension for `transb_` case, so I added it back.
The `opmath` is introduced to solve `Half/BFloat16`, so I guess I do not need to specializing `gemm_core` for Half/BFloat16 then.

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -263,11 +265,12 @@ scalar_t dot_naive(
+  opmath_t sum = 0;
   for (i = 0; i < n; i++) {
-    sum += op(x[i * incx], y[i * incy]);
+    sum += static_cast<opmath_t>(op(x[i * incx], y[i * incy]));
```
Shouldn't the cast happen before `op`

zhuhaozhe replied:
Thanks for advice, it will be better.

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -180,9 +181,10 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
   }
 
+  using opmath_t = at::opmath_type<scalar_t>;
   if ((trans == 'T') || (trans == 't')) {
```
What about the non-transposed case?

zhuhaozhe replied:
Thanks for reminder, have solved that too.

zhuhaozhe commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1550,18 +1551,20 @@ inline void baddbmm_cpu_kernel(const Tensor& result, const Tensor& self, const T
+            scalar_t& r = r2[j];
             if (is_bmm) {
-              r = 0;
+              acc_value = 0;
               for (const auto k : c10::irange(ks)) {
-                r += s2[k] * m1[k][j];
+                acc_value += s2[k] * m1[k][j];
               }
             } else {
-              r *= beta;
+              acc_value *= beta;
               for (const auto k : c10::irange(ks)) {
-                r += alpha * s2[k] * m1[k][j];
+                acc_value += alpha * s2[k] * m1[k][j];
               }
             }
+            r = acc_value;
```
Thanks for advice, I followed this with a little change since for baddbmm `acc_value` should got from `r2[j]`
```
opmath_t acc_value = is_bmm ? opmath_t(0) : opmath_t(r2[j]);
```

zhuhaozhe commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -53,32 +53,30 @@ auto sum(int64_t N, Func f) {
-        c[j * ldc + i_i * 4 + 1] += a[i_i * 4 + 1 + l * lda] * val;
-        c[j * ldc + i_i * 4 + 2] += a[i_i * 4 + 2 + l * lda] * val;
-        c[j * ldc + i_i * 4 + 3] += a[i_i * 4 + 3 + l * lda] * val;
+      const auto dot = sum(k, [&](int64_t l) -> opmath_t {
```
I use `sum` function in all for kernels now.

zhuhaozhe commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -323,8 +323,21 @@ void gemm(
 #endif
    gemm_stub(
-      at::kCPU, at::kBFloat16,
-      transa, transb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc);
+       at::kCPU,
+       at::kBFloat16,
+       transa,
+       transb,
+       m,
+       n,
+       k,
+       alpha,
+       a,
+       lda,
+       b,
+       ldb,
+       beta,
+       c,
+       ldc);
```
Will changed it back. It is auto changed by formatting, will do the formatting after review is done in the future.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -112,29 +111,29 @@ void gemm_transa_(
-        c[j * ldc + i_i * 4 + 3] += a[i_i * 4 + 3 + l * lda] * val;
+      const auto dot = sum(k, [&](int64_t l) -> opmath_t {
+        return static_cast<opmath_t>(a[l * lda + i]) *
+            static_cast<opmath_t>(b[l * ldb + j]);
```
The point of the old code is to access both arrays in contiguous order.

zhuhaozhe replied:
Yes, it is. But I guess if I follow that, I will need malloc larger temporary `c` buffer to hold `accumulation value`, please kindly correct me if I am wrong.
Do you think it is necessary to keep this optimization in the `reference path` and might introduce larger memory usage?

peterbell10 replied:
This is why I suggested specializing `gemm_core`. You can use the inefficient memory ordering for BFloat16 and Half, but the efficient ordering for all other types.

zhuhaozhe replied:
I specializing `gemm_notrans_` and `gemm_transb_` to keep efficient ordering for all other types.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
Hi, @mingfeima. I guess the remained failures https://hud.pytorch.org/pr/96074#11846003408 ( tiny difference between `spmm_reduce_kernel_impl` and `gemm_core_`) are caused by `spmm_reduce_kernel_impl` do not use `accumulate type` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/SpmmReduceKernel.cpp#L78-L87. I added precisionOverride on bf16(0.01) in `test_sparse_mm_reduce_sum`. And using `accumulate type` in `spmm_reduce_kernel_impl` should be another PR. Do you have some suggestions?

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -196,17 +198,24 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
   } else {
     if (beta != scalar_t(1) && beta != scalar_t(0)) scal<scalar_t>(m, beta, y, incy);
 
+    std::vector<opmath_t> sum(m);
+    for (const auto i : c10::irange(m)) {
+      sum[i] = y[i * incy];
+    }
```
You only need to use a buffer for BFloat16 and Half.

zhuhaozhe replied:
Sorry, I did not catch you up here.
From my understanding, only use a buffer will also have inefficient memory ordering like  in `gemm_core`, are you suggesting specializing `gemv` for BF16 with inefficient memory ordering and keep efficient memory ordering for other types. (I do not want to introduce Half in this PR, and there will be another PR to enable Half with this)

peterbell10 replied:
You don't need to change the iteration order, just accumulate in the result when `opmath_t` is the same as `scalar_t` (i.e. not BFloat16 and Half). 

>  do not want to introduce Half in this PR, and there will be another PR to enable Half with this

I don't understand your point here, this PR does enable opmath for Half...

zhuhaozhe replied:
Got it. I misunderstand "A" buffer before, I thought it means use a scalar buffer.
>   I don't understand your point here, this PR does enable opmath for Half...
I just saw some comments is mentioning `Half`. We can forget `Half` it is this PR, this should be another PR for `Half`

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1550,17 +1551,18 @@ inline void baddbmm_cpu_kernel(const Tensor& result, const Tensor& self, const T
-            scalar_t &r = r2[j];
+            opmath_t acc_value = 0;//is_bmm ? opmath_t(0) : opmath_t(r2[j]);
+            for (const auto k : c10::irange(ks)) {
+              acc_value += s2[k] * m1[k][j];
```
```suggestion
              acc_value += static_cast<opmath_t>(s2[k]) * static_cast<opmath_t>(m1[k][j]);
```

zhuhaozhe replied:
Done.
Just for curiosity, does this change is for readable only? The `bfloat16` type should have overrided `*` so the cast here should happen whether we add `static_cast<opmath_t>` or not.

peterbell10 replied:
`BFloat16::operator*` casts the result back to `BFloat16`, so this cast increases precision.

zhuhaozhe replied:
Got it, thanks.

peterbell10 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1550,17 +1551,18 @@ inline void baddbmm_cpu_kernel(const Tensor& result, const Tensor& self, const T
+              if (beta == scalar_t{0}) {
+                r2[j] = alpha * acc_value;
+              } else {
+                r2[j] = r2[j] * beta + alpha * acc_value;
```
The assignments of `alpha` and `beta` above should be changed to `.to<opmath_t>()`

zhuhaozhe replied:
Done.

peterbell10 commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -83,6 +83,36 @@ void gemm_notrans_(
+    int64_t m,
+    int64_t n,
+    int64_t k,
+    opmath_t alpha,
+    const scalar_t* a,
```
This doesn't need to be a template, and there's no need for `enable_if` anywhere in this file
```suggestion
    float alpha,
    const c10::BFloat16* a,
```

zhuhaozhe replied:
I used template with `enable_if` so the code changes will be smaller while supporting `Half`.

peterbell10 replied:
Okay, then why not do `is_same<scalar_t, opmath_t>`?

zhuhaozhe replied:
Good idea.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1550,17 +1551,20 @@ inline void baddbmm_cpu_kernel(const Tensor& result, const Tensor& self, const T
-              r = beta == scalar_t(0) ? scalar_t(0) : beta * r;
-              for (const auto k : c10::irange(ks)) {
-                r += alpha * s2[k] * m1[k][j];
+              if (beta == scalar_t{0}) {
+                r2[j] = static_cast<opmath_t>(alpha) * acc_value;
+              } else {
+                r2[j] = r2[j] * static_cast<opmath_t>(beta) +
+                    static_cast<opmath_t>(alpha) * acc_value;
```
No, you should change `.to<scalar_t>()` to `.to<opmath_t>()` above for beta and alpha. Otherwise you truncate the value.

zhuhaozhe replied:
Done.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -180,9 +181,10 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
   if ((trans == 'T') || (trans == 't')) {
     for (const auto i : c10::irange(n)) {
-      scalar_t sum = 0;
+      opmath_t sum = 0;
```
Just noticed `beta` and `alpha` are only `scalar_t` here. In future it would be nice for that to be `opmath_t` but it's okay as-is for this PR

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -196,15 +198,36 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
     if (beta != scalar_t(1) && beta != scalar_t(0)) scal<scalar_t>(m, beta, y, incy);
 
+    bool is_low_precision = !std::is_same<opmath_t, scalar_t>::value;
+    std::vector<opmath_t> sum;
+    if (is_low_precision) {
+      sum.reserve(m);
+      for (const auto i : c10::irange(m)) {
+        sum[i] = y[i * incy];
+      }
+    }
```
This is illegal, you've can't write to the vector with size 0. It would also make sense to avoid the copy here so we only access `y` once at the end.
```suggestion
    std::vector<opmath_t> sum;
    if (is_low_precision) {
      sum.resize(m);
    }
```

zhuhaozhe replied:
Thanks for reviewing, resolved.

peterbell10 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1,27 +1,27 @@
 #include <c10/util/irange.h>
 #include <c10/util/variant.h>
-
+#include <iostream>
```
```suggestion

```

zhuhaozhe replied:
thanks for the reminder, I removed.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -196,15 +198,33 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
+    }
     for (const auto j : c10::irange(n)) {
       scalar_t *column_ = a + lda * j;
       scalar_t z = alpha * x[j * incx];
```
```suggestion
      opmath_t z = alpha * static_cast<opmath_t>(x[j * incx]);
```

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -196,15 +198,33 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
         //output values are ignored if beta is 0, and set to 0, nans and infs are not propagated
         if (j==0 && beta==scalar_t(0)) {
-         y[i * incy] = scalar_t(0);
+          if (is_low_precision) {
+            sum[i] = 0;
+          } else {
```
```suggestion
          if (!is_low_precision) {
```

peterbell10 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -196,15 +198,33 @@ void gemv(char trans, int64_t m, int64_t n, scalar_t alpha, scalar_t *a, int64_t
+      }
+    }
+    if (is_low_precision) {
+      for (const auto i : c10::irange(m)) {
+        y[i * incy] = sum[i];
       }
```
You forgot to include the y value in the sum. I'm more than a little concerned this wasn't picked up in CI. Is this code not tested?

```suggestion
      if (beta == scalar_t(0) {
        for (const auto i : c10::irange(m)) {
          y[i * incy] = sum[i];
        }
      } else {
        for (const auto i : c10::irange(m)) {
          y[i * incy] += sum[i];
        }
      }
```

zhuhaozhe replied:
Yes you are correct. The test_addmv do not cover this reference path, I added ut to cover it.

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "use accmulate type in BF16 gemm(include dot, mv) ref path"

Fix https://github.com/pytorch/pytorch/issues/95125 and https://github.com/pytorch/pytorch/issues/83863 for bf16 accumulation in gemm ref path



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
